Fix so3 log and theta

nurlanov-zh opened the pull request:
The issue of optimizing pose-graph like problem using `ceres autodiff`, when the initial residual vector has an angle close to `\pi`, is addressed. 
The implementation of `logAndTheta` function in `so3.hpp` affects both SO3 and SE3 optimization problems. The new tests for SO3 and SE3 were added to show the issue of the previous implementation. There, `ceres autodiff` gives zero gradients wrt scalar part of a quaternion at the residuals having angle close to `\pi`.

To solve it, we propose to use a stable formula, when `w` is close to 0: 
` J.theta = Scalar(4) * atan(n / (w + sqrt(squared_w + squared_n))); `. 
In the above formula, we also utilize the property of quaternions: "`q` and `-q` represent the same rotation". So we can always choose the one with non-negative scalar part `w`, such that `w + sqrt(squared_w + squared_n)` is always `>= 1`.

Moreover, we derived an analytical formula for the jacobian of the logarithmic map `Dx_log_this()` with the above implementation and compared the analytical jacobian against automatically derived one by ceres in `test_ceres_so3.cpp`.

With the new implementation, we found the precision issue of `average` function for SO3 (`float`-valued). So we commented out tests comparing average vectors using `average` function against `interpolate` and `iaverage` functions.

strasdat commented:
@nurlanov-zh thanks for your contribution. I need to review this more carefully before I can merge this. You can help me by breaking this PR into smaller chunks. Starting with "SO3 log function is fixed for angles close to pi." and the corresponding test case would be a good start.

(I assume this is reacted to https://math.stackexchange.com/questions/83874/efficient-and-accurate-numerical-implementation-of-the-inverse-rodrigues-rotatio/110059#110059 which also noted that the case around pi is tricky.)

nurlanov-zh commented:
> @nurlanov-zh thanks for your contribution. I need to review this more carefully before I can merge this. You can help me by breaking this PR into smaller chunks. Starting with "SO3 log function is fixed for angles close to pi." and the corresponding test case would be a good start.
> 
@strasdat  Thanks for considering the PR. I have opened the [new smaller one](https://github.com/strasdat/Sophus/pull/281) with a fixed log map, where I also added corresponding test cases, and fixed some old tests that become not accurate enough with the new implementation.

The part I did not add from this PR to the new one is the analytical Jacobian of the SO3 log map, and its comparison against the automatically derived one.

> (I assume this is reacted to https://math.stackexchange.com/questions/83874/efficient-and-accurate-numerical-implementation-of-the-inverse-rodrigues-rotatio/110059#110059 which also noted that the case around pi is tricky.)

Yes, our concerns also started from the Rotation matrix to angle-axis transform around pi, however, we did not see this question in StackExchange before. Then we noted that the problem with zero gradients (around pi) occurs in the current implementation of the log map in Sophus.

strasdat commented:
Closing for now. Some parts of it seem to be addressed by https://github.com/strasdat/Sophus/pull/316.

Please reopen after resolving conflicts, in case there are some tests why might want to merge.
